core/src/main/java/com/uber/m3/tally/HistogramImpl.java

```diff
@@ -23,145 +23,144 @@
-        if (index >= buckets.size()) {
-            index = buckets.size() - 1;
-        }
-
-        buckets.get(index).samples.inc(1);
+        int index = toBucketIndex(Arrays.binarySearch(lookupByValue, value));
+        getOrCreateCounter(index).inc(1);
     }
 
     @Override
     public void recordDuration(Duration duration) {
-        int index = Collections.binarySearch(lookupByDuration, duration);
+        int index = toBucketIndex(Arrays.binarySearch(lookupByDuration, duration));
+        getOrCreateCounter(index).inc(1);
+    }
+
+    private CounterImpl getOrCreateCounter(int index) {
+        if (bucketCounters[index] != null) {
+            return bucketCounters[index];
+        }
+
+        // To maintain lock granularity we synchronize only on a
+        // particular bucket leveraging bucket's boundary as a sync target
+        synchronized (lookupByDuration[Math.min(index, lookupByDuration.length - 1)]) {
+            // Check whether bucket has been already set,
+            // while we were waiting for lock
+            if (bucketCounters[index] != null) {
+                return bucketCounters[index];
+            }
 
-        if (index < 0) {
-            // binarySearch returns the index of the search key if it is contained in the list;
-            // otherwise, (-(insertion point) - 1).
-            index = -(index + 1);
+            return (bucketCounters[index] = new HistogramBucketCounterImpl(scope, getQualifiedName(), index));
         }
+    }
 
-        // binarySearch can return collections.size(), guarding against that.
-        // pointing to last bucket is fine in that case because it's [_,infinity).
-        if (index >= buckets.size()) {
-            index = buckets.size() - 1;
+    private int toBucketIndex(int binarySearchResult) {
+        // Buckets are defined in the following way:
+        //      - Each bucket is inclusive of its lower bound, and exclusive of the upper: [lower, upper)
+        //      - All buckets are defined by upper bounds: [2, 4, 8, 16, 32, ...]: therefore i
+        //      in this case [-inf, 2) will be the first bucket, [2, 4) -- the second and so on
+        //
+        // Given that our buckets are designated as [lower, upper), and
+        // that the binary search is performed over upper bounds, if binary
+        // search found the exact match we need to shift it by 1 to index appropriate bucket in the
+        // array of (bucket's) counters
+        if (binarySearchResult >= 0) {
+            return binarySearchResult + 1;
         }
 
-        buckets.get(index).samples.inc(1);
+        // Otherwise, binary search will return {@code (-(insertion point) - 1)} where
+        // "insertion point" designates first element that is _greater_ than the key, therefore
+        // we simply use this an index in the array of counters
+        //
+        // NOTE: {@code ~binarySearchResult} is equivalent to {@code -(binarySearchResult) - 1}
+        return ~binarySearchResult;
```

Comment:
This hurts readability to save maybe 2-3 CPU cycles. Do you see any meaningful impact to your microbenchmarks with the simpler version?

Reply:
What exactly hurts readability?

Reply:
`~binarySearchResult` is less obvious than `-(binarySearchResult) - 1`. If it doesn't have any meaningful impact on performance, please use the simpler form.

Reply:
There's a comment right above it explaining it

Reply:
That's fair.